crates/transport/src/layers/retry.rs

```diff
@@ -0,0 +1,232 @@
+}
+
+impl RetryPolicy for RateLimitRetryPolicy {
+    fn should_retry(&self, error: &TransportError) -> bool {
+        error.is_retryable()
+    }
+
+    /// Provides a backoff hint if the error response contains it
+    fn backoff_hint(&self, error: &TransportError) -> Option<std::time::Duration> {
+        error.backoff_hint()
+    }
+}
+
+impl<S> Layer<S> for RetryBackoffLayer {
+    type Service = RetryBackoffService<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        RetryBackoffService {
+            inner,
+            policy: RateLimitRetryPolicy,
+            max_rate_limit_retries: self.max_rate_limit_retries,
+            max_timeout_retries: self.max_timeout_retries,
+            initial_backoff: self.initial_backoff,
+            compute_units_per_second: self.compute_units_per_second,
+            requests_enqueued: Arc::new(AtomicU32::new(0)),
+        }
+    }
+}
+
+/// A Tower Service used by the RetryBackoffLayer that is responsible for retrying requests based
+/// on the error type. See [TransportError] and [RateLimitRetryPolicy].
+#[derive(Debug, Clone)]
+pub struct RetryBackoffService<S> {
+    /// The inner service
+    inner: S,
+    /// The retry policy
+    policy: RateLimitRetryPolicy,
+    /// The maximum number of retries for rate limit errors
+    max_rate_limit_retries: u32,
+    /// The maximum number of retries for timeout errors
+    max_timeout_retries: u32,
+    /// The initial backoff in milliseconds
+    initial_backoff: u64,
+    /// The number of compute units per second for this service
+    compute_units_per_second: u64,
+    /// The number of requests currently enqueued
+    requests_enqueued: Arc<AtomicU32>,
+}
+
+impl<S> Service<RequestPacket> for RetryBackoffService<S>
+where
+    S: Service<RequestPacket, Response = ResponsePacket, Error = TransportError>
+        + Send
+        + 'static
+        + Clone,
+    S::Future: Send + 'static,
+{
+    type Response = ResponsePacket;
+    type Error = TransportError;
+    type Future = TransportFut<'static>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        // Our middleware doesn't care about backpressure, so it's ready as long
+        // as the inner service is ready.
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, request: RequestPacket) -> Self::Future {
+        let inner = self.inner.clone();
+        let this = self.clone();
+        let mut inner = std::mem::replace(&mut self.inner, inner);
+        Box::pin(async move {
+            let ahead_in_queue = this.requests_enqueued.fetch_add(1, Ordering::SeqCst) as u64;
+            let mut rate_limit_retry_number: u32 = 0;
+            let mut timeout_retries: u32 = 0;
+            loop {
+                let err;
+                let fut = inner.call(request.clone()).await;
+
+                match fut {
+                    Ok(res) => {
+                        if let Some(e) = res.as_error() {
```

Comment:
logic in this function should resolve any requests in the ResponsePacket that are resolvable, and should not re-try those requests

Reply:
this causes some problems with our batch model that need to be thought through carefully 🤔

Reply:
so i suppose when doing packets, the retry loop should 
- store successful responses
- retry ONLY the errors
- if retries fail, deliver all responses (successes and failures both)

Reply:
> logic in this function should resolve any requests in the ResponsePacket that are resolvable, and should not re-try those requests. As written, potentially fails to resolve some resolvable response

@prestwich Can you clarify here wdym by resolvable error?

Reply:
> so i suppose when doing packets, the retry loop should
> 
> * store successful responses
> * retry ONLY the errors
> * if retries fail, deliver all responses (successes and failures both)

Yeah, this works for batch responses.